### PR TITLE
[9.x.x] [ShaderGraph] Fixes for Virtual Texture Properties, Procedural VT, and Shader Preview

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR
Fixes a number of issues with VT:
* Preview correctly shows unassigned VT texture result (don't ignore null textures)
* Fix for duplicate layer naming
* Move TextureType to VTProperty, removed from old SampleVT node
* Fixed squished UI of VT property

---
### Testing status

Tested several VT use cases in editor and editor playmode:
VT enabled, PVT enabled - using PVT test project, and streaming VT
VT enabled - streaming VT
VT disabled - streaming VT
against latest 2020.1/staging

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/9.x.x%252Fsg%252Fvt-fixes

---

### CODE REVIEW

File: com.unity.shadergraph/Editor/Data/Graphs/SerializableVirtualTexture.cs [hunk 1]
```diff
@@ -4,37 +4,39 @@
 
 namespace UnityEditor.ShaderGraph.Internal
 {
+    // TODO: make TextureType work like this, combine the two enums (needs versioning for old data)
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Data/Graphs/SerializableVirtualTexture.cs:7]: really wary of shipping anymore "todo" code. can we do this now or make a better note to do it later? todo code almost never gets done

File: com.unity.shadergraph/Editor/Data/Graphs/VirtualTextureShaderProperty.cs [hunk 2]
```diff
@@ -108,10 +116,24 @@ internal override void AppendNonBatchablePropertyDeclarations(ShaderStringBuilde
                     builder.Append(value.layers[i].layerRefName);
                 }
                 builder.Append(")");
-                builder.AppendLine(delimiter);      // TODO: don't like delimiter, pretty sure it's not necessary if we invert the defaults on GEtPropertyDeclaration / GetPropertyArgument string
+                builder.Append(delimiter);      // TODO: don't like delimiter, pretty sure it's not necessary if we invert the defaults on GEtPropertyDeclaration / GetPropertyArgument string
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Data/Graphs/VirtualTextureShaderProperty.cs:119]: same as above

File: com.unity.shadergraph/Editor/Data/Nodes/AbstractMaterialNode.cs [hunk 3]
```diff
@@ -407,6 +407,13 @@ public AbstractShaderProperty GetSlotProperty(int inputSlotId)
...
                 }
 
+#if PROCEDURAL_VT_IN_GRAPH
+                if (fromNode is ProceduralVirtualTextureNode pvtNode)
+                {
+                    return pvtNode.AsShaderProperty();
+                }
+#endif // PROCEDURAL_VT_IN_GRAPH
+
                 // TODO: what if it's from a subgraph?  should probably disallow that
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Data/Nodes/AbstractMaterialNode.cs:417]: same as above

File: com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/ProceduralVirtualTextureNode.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+        const string kOutputSlotName = "Out";
+
+        public ProceduralVirtualTextureNode()
+        {
+            UpdateNodeAfterDeserialization();
+            SetLayerCount(2);
+
+            vtProperty.displayName = "ProceduralVirtualTexture";
+            vtProperty.overrideReferenceName = "MyPVT";             // TODO : make unique
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/ProceduralVirtualTextureNode.cs:23]: same as above

File: com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/ProceduralVirtualTextureNode.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        public int layers
+        {
+            get { return vtProperty.value.layers.Count; }
+            set
+            {
+                if (vtProperty.value.layers.Count == value)
+                    return;
+                // TODO: is there an easier way to do this ?
+                SetLayerCount(value);
+                Dirty(ModificationScope.Topological);       // TODO: need some way to re-run node setup on dependent nodes, so sampleVT nodes can update their output count
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/ProceduralVirtualTextureNode.cs:80]: same as above, this one Liz might be able to help with

File: com.unity.shadergraph/Editor/Data/Graphs/SerializableVirtualTexture.cs (hunk 1 above)
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Data/Graphs/SerializableVirtualTexture.cs:7]: I can move it to a feature request in fogbugz instead.   Definitely do not want to make that change in this PR....  :P

**Comment by cdxntchou** [comment]
Yamato green except for broken tests (OSX license downloader)

**Pull request merged**